kt/godot-runtime/src/main/kotlin/godot/core/GarbageCollector.kt

```diff
@@ -0,0 +1,143 @@
+
+    private var forceJvmGarbageCollector = false
+    private var gcState = GCState.NONE
+    val isClosed: Boolean
+        get() = gcState == GCState.CLOSED
+
+    fun registerInstance(instance: KtObject) {
+        synchronized(this) {
+            val rawPtr = instance.rawPtr
+            if (instance.isRef) {
+                refWrappedMap[rawPtr] = WeakReference(instance)
+                MemoryBridge.ref(rawPtr)
+            } else {
+                wrappedMap[rawPtr] = instance
+            }
+        }
+    }
+
+    fun getObjectInstance(ptr: VoidPtr): KtObject? {
+        val ktObject = wrappedMap[ptr]
+        return if (ktObject != null) {
+            if (MemoryBridge.checkInstance(ptr, ktObject.godotInstanceId)) ktObject else null
+        } else null
+    }
+
+    fun start(forceJvmGarbageCollector: Boolean) {
+        this.forceJvmGarbageCollector = forceJvmGarbageCollector
+        start()
+        gcState = GCState.STARTED
+    }
+
+    fun close() {
+        gcState = GCState.CLOSED
+        println("Closing GC thread ...")
+    }
+
+    fun cleanUp() {
+        val begin = Instant.now()
+        while (refWrappedMap.isNotEmpty() || wrappedMap.isNotEmpty()) {
+            forceJvmGc()
+            checkAndClean()
+            val finish = Instant.now()
+            if (Duration.between(begin, finish).toMillis() > 5000) {
+                throw GCEndException(
+                        buildString {
+                            append("Some JVM godot instances are leaked.")
+                            append(System.lineSeparator())
+                            append("Leaked references:")
+                            append(System.lineSeparator())
+                            for (entry in refWrappedMap) {
+                                append("    ${entry.key}: ${entry.value}")
+                                append(System.lineSeparator())
+                            }
+                            append("Leaked objects:")
+                            append(System.lineSeparator())
+                            for (entry in wrappedMap) {
+                                append("    ${entry.key}: ${entry.value}")
+                                append(System.lineSeparator())
+                            }
+                        }
+                )
+            }
+        }
+
+    }
+
+    override fun run() {
+        while (gcState == GCState.STARTED) {
+            if (forceJvmGarbageCollector) {
+                forceJvmGc()
+            }
+            checkAndClean()
+            sleep(500)
+        }
+        gcState = GCState.CLOSED
+    }
+
+    private fun checkAndClean() {
+        // A native reference cannot die while a jvm instance exists (because counter > 0). When we don't need the
+        // jvm instance anymore, we decrease the counter.
+        for (entry in refWrappedMap) {
+            if (entry.value.get() == null) {
+                if (MemoryBridge.unref(entry.key)) {

```

Comment:
something for later: we probably can batch the call here to lessen the amount of calls we make to cpp.

Reply:
Yep, that was one of our consideration with @CedNaru